### PR TITLE
Inline attachments

### PR DESCRIPTION
**Pull request by theodowling**
Hi there

I needed inline-attachment support for a project, so I added it for myself, and thought it might make a good PR.

**Inline-Attachment support has been added to Sendgrid, Mandrill, Mailgun, Postmark and SparkPost adapters**.

Update: SMTP adapter also supported

At its core you just need to specify `type: :inline` when creating an Attachment, and all the rest will happen automatically.

```
Attachment.new("/data/file.png", type: :inline)
```

And the ContentID of the attachment will default to filename so in the above case, you can refer to the inline-attached image by using:
```
<img src="cid:file.png">
```

I've tested using all of the adapters listed above. Please let me know if you have any comments, or need me to change anything.

Kind regards
Theo

### CODE REVIEW

**Comment by princemaple** [comment]
Hi Theo,

I'm on my phone.
I just want to quickly say thank you for the great effort!

I've scanned through the PR, it's mostly good. I'll go through it again on my desktop soon. Probably just nits so I might simply merge this and tweak the styles / implementation details myself.

Po

File: lib/swoosh/adapters/mailgun.ex [hunk 1]
```diff
@@ -86,10 +88,23 @@ defmodule Swoosh.Adapters.Mailgun do
 
   defp prepare_attachments(body, %{attachments: []}), do: body
   defp prepare_attachments(body, %{attachments: attachments}) do
-    Map.put(body, :attachments, Enum.map(attachments, &prepare_file(&1)))
+    normal_attachments = Enum.filter(attachments, fn(%Swoosh.Attachment{type: type}) -> type == :attachment end)
+    inline_attachments = Enum.filter(attachments, fn(%Swoosh.Attachment{type: type}) -> type == :inline end)
```
Comment (princemaple) [inline lib/swoosh/adapters/mailgun.ex]: - `Enum.split_with`
- Matching struct type in private functions is not necessary

File: lib/swoosh/adapters/mailgun.ex [hunk 2]
```diff
@@ -118,12 +133,15 @@ defmodule Swoosh.Adapters.Mailgun do
   defp prepare_html(body, %{html_body: nil}), do: body
   defp prepare_html(body, %{html_body: html_body}), do: Map.put(body, :html, html_body)
 
-  defp encode_body(%{attachments: attachments} = params) do
+  defp encode_body(%{attachments: attachments, inline: inline} = params) do
     {:multipart,
      params
      |> Map.drop([:attachments])
+     |> Map.drop([:inline])
```
Comment (princemaple) [inline lib/swoosh/adapters/mailgun.ex]: `Map.drop([:attachments, :inline])`

File: lib/swoosh/adapters/mandrill.ex [hunk 3]
```diff
@@ -94,11 +94,22 @@ defmodule Swoosh.Adapters.Mandrill do
...
+    inline_attachments = Enum.filter(attachments, fn(attachment) -> attachment.type == :inline end)
+    
+    normal_attachments = Enum.map(normal_attachments, fn %Swoosh.Attachment{content_type: type, path: path, filename: filename} ->
+      content = path |> File.read! |> Base.encode64
+      %{type: type, name: filename, content: content}
+    end)
+
+    inline_attachments = Enum.map(inline_attachments, fn %Swoosh.Attachment{content_type: type, path: path, filename: filename} ->
+      content = path |> File.read! |> Base.encode64
+      %{type: type, name: "#{filename}", content: content}
```
Comment (princemaple) [inline lib/swoosh/adapters/mandrill.ex]: - I'd extract this common part into a separate function
- `"#{filename}"` is just `filename`?

File: lib/swoosh/attachment.ex [hunk 4]
```diff
@@ -3,7 +3,7 @@ defmodule Swoosh.Attachment do
   Struct representing an attachment in an email.
   """
 
-  defstruct filename: nil, content_type: nil, path: nil
+  defstruct filename: nil, content_type: nil, path: nil, type: nil
```
Comment (princemaple) [inline lib/swoosh/attachment.ex]: - Maybe just `inline: false` (default to attachment)?
- Should we do something special about inline attachments? e.g. validating it's an image?

File: lib/swoosh/email.ex [hunk 5]
```diff
@@ -464,23 +464,30 @@ defmodule Swoosh.Email do
   as an argument. If you give a path we will detect the MIME type and determine the filename
   automatically.
 
+  You can also send an inline-attachment used for embedding images in the body of emails by specifying `type: "inline"`
```
Comment (princemaple) [inline lib/swoosh/email.ex]: Currently it's `:inline` not `"inline"`
(might change, see comment above)

**Comment by theodowling** [comment]
Thanks @princemaple - I'll wait for you and @stevedomin to discuss the handling of `type: :inline|:attachment` vs `inline: true|false`. I myself wasn't sure which approach to take.

I'll also commit some small fixes/changes on the back of your comments above.

I haven't looked at the SMTP implementation, but I can do that also today.

File: lib/swoosh/adapters/mailgun.ex (hunk 2 above)
Comment (theodowling) [reply lib/swoosh/adapters/mailgun.ex]: Thanks. I'll update this.

File: lib/swoosh/adapters/mailgun.ex (hunk 1 above)
Comment (theodowling) [reply lib/swoosh/adapters/mailgun.ex]: Agreed. Thanks for recommending `Enum.split_with`. I'll fix this on my side.

File: lib/swoosh/adapters/mandrill.ex (hunk 3 above)
Comment (theodowling) [reply lib/swoosh/adapters/mandrill.ex]: This has also been updated. Thanks.

File: lib/swoosh/attachment.ex (hunk 4 above)
Comment (stevedomin) [reply lib/swoosh/attachment.ex]: I'm ok with using `type`. Slightly more future-proof.

@princemaple: regarding validation of inline attachments, do you know if the spec says anything about accepted file type for inline attachments?

File: mix.lock [hunk 6]
```diff
@@ -1,20 +1,20 @@
-%{"bunt": {:hex, :bunt, "0.2.0", "951c6e801e8b1d2cbe58ebbd3e616a869061ddadcc4863d0a2182541acae9a38", [:mix], [], "hexpm"},
```
Comment (stevedomin) [inline mix.lock]: New version of Hex?

File: lib/swoosh/email.ex (hunk 5 above)
Comment (stevedomin) [reply lib/swoosh/email.ex]: 👍

File: mix.lock (hunk 6 above)
Comment (princemaple) [reply mix.lock]: More like an older version. @theodowling Please exclude the mix.lock change :) You might want to upgrade with `mix local.hex`

Comment (theodowling) [reply mix.lock]: Thanks for the heads-up. I've reverted change to `mix.lock`

**Comment by theodowling** [comment]
Thanks for the Approval. 
I'm still working on adding inline-attachment support to SMTP adapter. Will push that as soon as I get it working 😄

File: lib/swoosh/attachment.ex (hunk 4 above)
Comment (princemaple) [reply lib/swoosh/attachment.ex]: I actually failed to search for the inline attachment related specs. :(

**Comment by princemaple** [comment]
@theodowling thanks for looking into it!

**Comment by stevedomin** [comment]
@theodowling nice work on the SMTP attachment!

File: lib/swoosh/adapters/mailgun.ex [hunk 7]
```diff
@@ -86,13 +88,17 @@ defmodule Swoosh.Adapters.Mailgun do
 
   defp prepare_attachments(body, %{attachments: []}), do: body
   defp prepare_attachments(body, %{attachments: attachments}) do
-    Map.put(body, :attachments, Enum.map(attachments, &prepare_file(&1)))
+    {normal_attachments, inline_attachments} = Enum.split_with(attachments, fn(%Swoosh.Attachment{type: type}) -> type == :attachment end)
```
Comment (princemaple) [inline lib/swoosh/adapters/mailgun.ex]: separate lines
```
{..} =
  ...
```

File: lib/swoosh/adapters/mandrill.ex [hunk 8]
```diff
@@ -94,11 +94,19 @@ defmodule Swoosh.Adapters.Mandrill do
 
   defp prepare_attachments(body, %{attachments: []}), do: body
   defp prepare_attachments(body, %{attachments: attachments}) do
-    Map.put(body, "attachments", Enum.map(attachments, &%{
-      "name" => &1.filename,
-      "type" => &1.content_type,
-      "content" => &1.path |> File.read! |> Base.encode64
-    }))
+    normal_attachments = prepare_attachments_structure(Enum.filter(attachments, fn(attachment) -> attachment.type == :attachment end))
+    inline_attachments = prepare_attachments_structure(Enum.filter(attachments, fn(attachment) -> attachment.type == :inline end))
```
Comment (princemaple) [inline lib/swoosh/adapters/mandrill.ex]: `Enum.split_with` (might as well just copy paste the change from mailgun)

File: lib/swoosh/adapters/mailgun.ex (hunk 7 above)
Comment (princemaple) [reply lib/swoosh/adapters/mailgun.ex]: Remove `Swoosh.Attachment` and unnecessary braces, `fn %{...} -> ... end is sufficient`

File: lib/swoosh/adapters/mandrill.ex [hunk 9]
```diff
@@ -94,11 +94,19 @@ defmodule Swoosh.Adapters.Mandrill do
...
+
+    body
+    |> Map.put("attachments", normal_attachments)
+    |> Map.put("images", inline_attachments)
+  end
+
+  defp prepare_attachments_structure(attachments) do
+    Enum.map(attachments, fn %Swoosh.Attachment{content_type: type, path: path, filename: filename} ->
```
Comment (princemaple) [inline lib/swoosh/adapters/mandrill.ex]: remove `Swoosh.Attachment`

File: lib/swoosh/adapters/postmark.ex [hunk 10]
```diff
@@ -81,11 +81,19 @@ defmodule Swoosh.Adapters.Postmark do
 
   defp prepare_attachments(body, %{attachments: []}), do: body
   defp prepare_attachments(body, %{attachments: attachments}) do
-    Map.put(body, "Attachments", Enum.map(attachments, &%{
-      "Name" => &1.filename,
-      "ContentType" => &1.content_type,
-      "Content" => &1.path |> File.read! |> Base.encode64
-    }))
+    Map.put(body, "Attachments", Enum.map(attachments, fn(attachment) -> prepare_attachment(attachment) end))
```
Comment (princemaple) [inline lib/swoosh/adapters/postmark.ex]: Remove unnecessary braces in fn arguments

File: lib/swoosh/adapters/postmark.ex [hunk 11]
```diff
@@ -81,11 +81,19 @@ defmodule Swoosh.Adapters.Postmark do
...
+  end
+
+  defp prepare_attachment(attachment) do
+    case attachment.type do
+      :inline -> %{"Name" => attachment.filename,
+                    "ContentType" => attachment.content_type,
```
Comment (princemaple) [inline lib/swoosh/adapters/postmark.ex]: indentation is off

File: lib/swoosh/adapters/sendgrid.ex [hunk 12]
```diff
@@ -106,9 +106,12 @@ defmodule Swoosh.Adapters.Sendgrid do
 
   defp prepare_attachments(body, %{attachments: []}), do: body
   defp prepare_attachments(body, %{attachments: attachments}) do
-    attachments = Enum.map(attachments, fn %{content_type: type, path: path, filename: filename} ->
+    attachments = Enum.map(attachments, fn %Swoosh.Attachment{content_type: content_type, path: path, filename: filename, type: type} ->
```
Comment (princemaple) [inline lib/swoosh/adapters/sendgrid.ex]: Remove `Swoosh.Attachment`

File: lib/swoosh/adapters/sparkpost.ex [hunk 13]
```diff
@@ -57,6 +57,8 @@ defmodule Swoosh.Adapters.SparkPost do
     html_body: html,
     attachments: attachments
   } = email) do
+    normal_attachments = Enum.filter(attachments, fn(attachment) -> attachment.type == :attachment end)
+    inline_attachments = Enum.filter(attachments, fn(attachment) -> attachment.type == :inline end)
```
Comment (princemaple) [inline lib/swoosh/adapters/sparkpost.ex]: split_with

**Comment by theodowling** [comment]
Hi @princemaple, thanks for the comments. I've made the changes as requested.
Please let me know if there is anything else we need to update/change. I'm happy to help out 👍

**Comment by stevedomin** [comment]
I'm 👍, feel free to merge if you feel it's good to go @princemaple

**Comment by princemaple** [comment]
@theodowling thank you very much for the great work. Merging now!

**Comment by blackham** [comment]
You would do well to update your documentation to mention the bit about the cid name being the filename. Even better to provide a simple example. (searching github bugs in a poor replacement for docs)

**Comment by princemaple** [comment]
@blackham It's not very helpful to comment on a PR closed over 3 years ago.

It'd be better you filed an issue to tell us what troubled you, and how we can improve. Basically what you did with the comment. Even better you could send a PR to address the issue.

**Comment by blackham** [comment]
Sorry it was late. Let me try and explain and give you example code you might be able to use in the doc. Also sorry for commenting on an old bug. 

It would have saved me some time last night if the text "Inline attachments by default use their filename (or basename of the path if filename is not specified) as cid, in relevant adapters." found in https://hexdocs.pm/swoosh/Swoosh.Attachment.html also gave a very basic example. 

Just this would have been enough for the light bulb to go off. 
``

Or better, I could have skipped most everything else in the docs and just looked at this example: 
```
   new()
    |> to({user.name, user.email})
    |> from({"Red Skull", "red_skull@badguyrus.org"})
    |> subject("OTP QR Code")
    |> html_body("<h1>Hello #{user.name}</h1> Here is your QR Code <img src='cid:qrcode.png'>")
    |> text_body("Hello #{user.name}. Please find your QR Code attached.\n")
    |> attachment(
      Swoosh.Attachment.new(
        {:data, Avengers.Accounts.OTP.qrcode_png(user)},                                                                                     
        filename: "qrcode.png",
        content_type: "image/png",
        type: :inline)
    )
    |> Avengers.Mailer.deliver

```

Maybe OTP (One Time Password) isn't the best in that example. Using "OTP" for anything but Open Telecom Platform in Elixir might be against some peoples religion.  But TwoFA.qrcode_png(user) is ugly. TwoFactorAuth.qrcode_png(user) or TFA.qrcode_png(user)? In any case seeing the  is where my oh yeah, I'm stupid kicked in. 

BTW, Great job on a mail system!

**Comment by princemaple** [comment]
Thanks for that. I'll update the docs to include your example.

Documentation PRs are always welcome.

**Comment by princemaple** [comment]
Pushed more docs updates.

**Pull request merged**